ouroboros-consensus/src/Ouroboros/Consensus/Storage/ChainDB/Impl/VolDB.hs

```diff
@@ -318,56 +546,90 @@ isReachable predecessor i b =
   -> StreamFrom blk
   -> StreamTo   blk
   -> Maybe (Path blk)
-computePath predecessor from to = case to of
-    StreamToInclusive (RealPoint _ end)
-      | Just prev <- predecessor end -> case from of
-          -- Easier to handle this special case (@StreamFromInclusive start,
-          -- StreamToInclusive end, start == end@) here:
-          StreamFromInclusive (RealPoint _ start)
-            | start == end -> return $ CompletelyInVolDB [end]
-          _                -> go [end] prev
-      | otherwise        -> return $ NotInVolDB end
-    StreamToExclusive (RealPoint _ end)
-      | Just prev <- predecessor end -> go [] prev
-      | otherwise                    -> return $ NotInVolDB end
+computePath lookupBlockInfo from to =
+    case computeReversePath lookupBlockInfo (realPointHash endPt) of
+      Nothing
+        -> Just $ NotInVolDB endPt
+      Just volPath
+        | exclUpperBound
+        -- When there's an exclusive upper bound, we must exclude the first
+        -- point of the reverse path from the returned 'Path'. We simply drop
+        -- it from the path before starting the main loop (@go@).
+        -> case volPath of
+             volPath' ::> _   -> go [] volPath'
+             -- If we're immediately stopped, the accumulator will be empty,
+             -- so no need to drop the last element. Reuse the termination
+             -- logic of @go@.
+             StoppedAt {}     -> go [] volPath
+             -- The exclusive end bound cannot be genesis, as it must be a
+             -- 'RealPoint', so the reverse path cannot start with
+             -- 'StoppedAtGenesis'.
+             StoppedAtGenesis -> error "exclusive end bound cannot be genesis"
+        | otherwise
+        -> go [] volPath
   where
-    -- | It only needs the predecessor @prev@ and not the actual hash, because
-    -- the hash is already added to @acc@ (if allowed by the bounds).
-    go :: [HeaderHash blk] -> WithOrigin (HeaderHash blk) -> Maybe (Path blk)
-    go acc prev = case prev of
-      -- Found genesis
-      Origin -> case from of
-        StreamFromInclusive _
-          -> Nothing
-        StreamFromExclusive GenesisPoint
-          -> return $ CompletelyInVolDB acc
-        StreamFromExclusive (BlockPoint {})
+    (endPt, exclUpperBound) = case to of
+        StreamToInclusive pt -> (pt, False)
+        StreamToExclusive pt -> (pt, True)
+
+    fieldsToRealPoint :: HeaderFields blk -> RealPoint blk
+    fieldsToRealPoint flds =
+        RealPoint (headerFieldSlot flds) (headerFieldHash flds)
+
+    -- | Convert the 'HeaderFields' to a 'RealPoint' and prepend that to the
+    -- accumulator.
+    --
+    -- NOTE: forces the 'RealPoint' to avoid holding onto the lazy
+    -- 'HeaderFields' data type, which is derived from a 'VolDB.BlockInfo'
+    -- (see 'computeReversePath'), which is much larger than a 'RealPoint'.
+    addToAcc :: HeaderFields blk -> [RealPoint blk] -> [RealPoint blk]
+    addToAcc flds = (pt :)
```

Comment:
Eta-expand and add comment: if this list is forced, that forces `pt`, and the list is forced because the accumulator is forced in `go`.